src/backend/distributed/executor/local_executor.c

```diff
@@ -0,0 +1,579 @@
+
+		totalRowsProcessed += ExecuteLocalTaskPlan(node, localPlan, task->queryString);
+	}
+
+	return totalRowsProcessed;
+}
+
+
+/*
+ * GetParamListOids gets a ParamListInfo and returns an array of Oids
+ * where each element is the Oid of the corresponding Param in the ParamListInfo.
+ *
+ * If paramList is NULL or doesn't have any parameters, the function returns
+ * NULL.
+ */
+static Oid *
+GetParamListOids(ParamListInfo paramList)
+{
+	Oid *paramListTypes = NULL;
+	int numParamIndex = 0;
+
+	if (paramList == NULL || paramList->numParams <= 0)
+	{
+		return NULL;
+	}
+
+	paramListTypes = (Oid *) palloc(sizeof(Oid) * paramList->numParams);
+
+	for (numParamIndex = 0; numParamIndex < paramList->numParams; numParamIndex++)
+	{
+		ParamExternData *param;
+
+		/* give hook a chance in case parameter is dynamic */
+		if (paramList->paramFetch != NULL)
+		{
+#if (PG_VERSION_NUM >= 110000)
+			ParamExternData paramData;
+
+			param =
+				paramList->paramFetch(paramList, numParamIndex + 1, false, &paramData);
+#else
+			param = &paramList->params[numParamIndex];
+			if (!OidIsValid(param->ptype))
+			{
+				paramList->paramFetch(paramList, numParamIndex + 1);
+			}
+#endif
+		}
+		else
+		{
+			param = &paramList->params[numParamIndex];
+		}
+
+		if (param->isnull)
+		{
+			paramListTypes[numParamIndex] = InvalidOid;
+			continue;
+		}
+
+		paramListTypes[numParamIndex] = param->ptype;
+	}
+
+	return paramListTypes;
+}
+
+
+/*
+ * ExecuteLocalTaskPlan gets a planned statement which can be executed locally.
+ * The function simply follows the steps to have a local execution, sets the
+ * tupleStore if necessary. The function returns the
+ */
+static uint64
+ExecuteLocalTaskPlan(CitusScanState *scanState, PlannedStmt *taskPlan, char *queryString)
+{
+	EState *executorState = ScanStateGetExecutorState(scanState);
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+	DestReceiver *tupleStoreDestReceiever = CreateDestReceiver(DestTuplestore);
+	ScanDirection scanDirection = ForwardScanDirection;
+	QueryEnvironment *queryEnv = create_queryEnv();
+	QueryDesc *queryDesc = NULL;
+	int eflags = 0;
+	uint64 totalRowsProcessed = 0;
+
+	/*
+	 * Use the tupleStore provided by the scanState because it is shared accross
+	 * the other task executions and the adaptive executor.
+	 */
+	SetTuplestoreDestReceiverParams(tupleStoreDestReceiever,
+									scanState->tuplestorestate,
+									CurrentMemoryContext, false);
+
+	/* Create a QueryDesc for the query */
+	queryDesc = CreateQueryDesc(taskPlan, queryString,
+								GetActiveSnapshot(), InvalidSnapshot,
+								tupleStoreDestReceiever, paramListInfo,
+								queryEnv, 0);
+
+	ExecutorStart(queryDesc, eflags);
+	ExecutorRun(queryDesc, scanDirection, 0L, true);
+
+	/*
+	 * We'll set the executorState->es_processed later, for now only remember
+	 * the count.
+	 */
+	if (taskPlan->commandType != CMD_SELECT)
+	{
+		totalRowsProcessed = queryDesc->estate->es_processed;
```

Comment:
I think the reason you didn't reuse `ExecutePlanIntoDestReceiver` is to get the number of rows, though maybe `ExecutePlanIntoDestReceiver` can return the number using portal->queryDesc?

Reply:
That's a good idea, tried with it but failed. Postgres seem set queryDesc not for all types of [queries](https://github.com/postgres/postgres/blob/eb43f3d19324d7e5376b1f57fc2e5c142a6b5f3d/src/backend/tcop/pquery.c#L482-L537). So, I'm skipping this suggestion?